File: tools/thesis_main/analysis/image_portrait/review_handoff.js
/* Optional human-review panel for the existing Studio. No annotation writeback. */
(() => {
 const d=window.DIFFICULTY_REVIEW;if(!d)return;
 const key='image-difficulty-review-20260915-v1',schema='image_difficulty_user_review_v1';
 const rows=new Map(d.rows.map(r=>[r.image_id,r])),units=new Map(d.rows.flatMap(r=>r.units.map(u=>[r.image_id+'|'+u.condition,u])));
 const E=s=>String(s??'').replace(/[&<>"']/g,c=>({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
 const base=()=>({status:'未审核',grade:'',reason:'',cluster_semantics:'',singleton_assessment:''});
 const names={simple:'简单',simple_candidate:'简单候选',medium:'中等',medium_candidate:'中等结构候选',difficult_candidate:'困难候选'};
 let edits={},notes={},blocked=false,message='';
 function validate(v){
  if(!v||!['未审核','暂缓','已审核'].includes(v.status)||!['','简单','中等','困难候选','三档不适用'].includes(v.grade)||(v.status==='已审核')!==!!v.grade)throw Error('审核状态与类别不匹配');
  for(const k of ['reason','cluster_semantics','singleton_assessment'])if(typeof v[k]!=='string')throw Error('缺少审核文本');
 }
 function validateImport(v){
  if(v.schema!==schema||v.source_v2!==d.source_v2||!Array.isArray(v.reviews)||!Array.isArray(v.image_notes))throw Error('审核版本不匹配');
  const a={},b={};for(const r of v.reviews){const k=r.image_id+'|'+r.condition;if(!units.has(k)||Object.hasOwn(a,k))throw Error('图片条件重复或未知');validate(r.review);a[k]=r.review;}
  for(const r of v.image_notes){if(!rows.has(r.image_id)||Object.hasOwn(b,r.image_id)||typeof r.note!=='string')throw Error('图片笔记无效');b[r.image_id]=r.note;}
  if(Object.keys(a).length!==units.size||Object.keys(b).length!==rows.size)throw Error('导出记录不完整');return {edits:a,notes:b};
 }
 function exportData(){return {schema,source_v1:d.source_v1,source_v2:d.source_v2,saved_at:new Date().toISOString(),reviews:[...units.keys()].map(k=>{const [image_id,condition]=k.split('|');return {image_id,condition,review:edits[k]||base()};}),image_notes:d.rows.map(r=>({image_id:r.image_id,note:notes[r.image_id]||''}))};}
 try{const v=localStorage.getItem(key);if(v)({edits,notes}=validateImport(JSON.parse(v)));}catch(e){blocked=true;message='旧缓存异常，暂停覆盖；请导入备份。'+e.message;}
 const panel=document.createElement('section');panel.className='history-panel';panel.innerHTML=`<h2>图片难度审核</h2><p>两版算法候选、原106张tag、你的审核分别保存。先看原图，再展开真人证据。拟合更规整不代表更正确；未审核不等于认可候选。</p><div class="history-controls"><label>图片搜索 <input id="rv-search" placeholder="楼名 / 图号 / 场景"></label><select id="rv-image" aria-label="审核图片"></select><button id="rv-prev">上一张</button><button id="rv-next">下一张</button><button id="rv-export">导出审核JSON</button><label>导入 <input id="rv-import" type="file" accept=".json"></label></div><p id="rv-status" role="status"></p><p id="rv-save" role="status"></p><details open><summary>原图（不叠加标注）</summary><a id="rv-original-link" target="_blank"><img id="rv-original" style="width:100%;height:auto" alt="当前图片原图"></a></details><label>仅根据可见图像的观察<textarea id="rv-note" rows="2" style="width:100%" placeholder="具体边界、遮挡位置、开口与反射；未知不猜测"></textarea></label><details><summary>展开候选分类及你的审核</summary><select id="rv-mode" aria-label="审核条件"></select><div id="rv-evidence"></div><label>你的裁决 <select id="rv-decision"><option>未审核</option><option>暂缓</option><option>简单</option><option>中等</option><option>困难候选</option><option>三档不适用</option></select></label><label>理由<textarea id="rv-reason" rows="2" style="width:100%"></textarea></label><div class="history-controls"><label>簇的实际含义 <input id="rv-cluster"></label><label>单人标法判断 <input id="rv-singleton"></label></div></details>`;
 document.querySelector('main').prepend(panel);const $r=id=>document.getElementById('rv-'+id);
 const opt=(v,t)=>{const x=document.createElement('option');x.value=v;x.textContent=t;return x;};
 function progress(){$r('status').textContent=`276图／230个图×条件；已审核 ${Object.values(edits).filter(v=>v.status==='已审核').length}，暂缓 ${Object.values(edits).filter(v=>v.status==='暂缓').length}。`;}
 function save(){progress();if(blocked)return;try{localStorage.setItem(key,JSON.stringify(exportData()));$r('save').textContent='已在浏览器保存，请定期导出JSON。';}catch(e){$r('save').textContent='自动保存失败，请立即导出：'+e.message;}}
 function mode(){const r=rows.get(dataset.cases[currentCase].image_id),c=$r('mode').value,u=r.units.find(x=>x.condition===c),v=edits[r.image_id+'|'+c]||base();
  for(const k of ['decision','reason','cluster','singleton'])$r(k).disabled=!u;
  $r('decision').value=v.status==='已审核'?v.grade:v.status;$r('reason').value=v.reason;$r('cluster').value=v.cluster_semantics;$r('singleton').value=v.singleton_assessment;
  $r('evidence').innerHTML=`<p>原tag：${E(r.expert?.expert_tag||'无')}；${E(r.scene)}</p>`+(u?`<p>第一轮：${E(names[u.previous.grade]||'未分类')}；第二轮：<b>${E(names[u.candidate['建议粗类_非最终']]||u.candidate['建议粗类_非最终']||'未分类')}</b></p><p>${['原作答人数','有效独立人数','单列无效人数','各簇人数_含单人','支持簇数','单人簇数','前8人完整后缀比例','至观察窗口支持核心稳定比例'].map(k=>E(k)+'：'+E(u.candidate[k]||'未计算')).join('；')}</p><details><summary>完整原始数值依据</summary><pre style="white-space:pre-wrap">${E(JSON.stringify(u,null,2))}</pre></details>`:'<p>无本轮历史，不生成历史粗类。可以填写图片观察。</p>');
  if(u){const s=document.getElementById('history-mode');s.value=c==='manual'?'Manual':'Semi';s.dispatchEvent(new Event('change'));}
 }
 function render(){const r=rows.get(dataset.cases[currentCase].image_id);if(!r)return;$r('image').value=r.image_id;$r('mode').replaceChildren(...r.units.map(u=>opt(u.condition,u.condition)));$r('note').value=notes[r.image_id]||'';$r('original').src='../../../'+r.image_path;$r('original-link').href='../../../'+r.image_path;mode();progress();}
 function choices(){const q=$r('search').value.toLowerCase();$r('image').replaceChildren(...d.rows.filter(r=>[r.image_id,r.number,r.scene].join(' ').toLowerCase().includes(q)).map(r=>opt(r.image_id,`${r.priority?'优先 · ':''}${r.building} · ${r.number} · ${r.scene}`)));}
 function open(){const i=dataset.cases.findIndex(c=>c.image_id===$r('image').value);if(i>=0)chooseCase(i);}
 $r('search').oninput=choices;$r('image').onchange=open;$r('mode').onchange=mode;
 for(const [k,n] of [['prev',-1],['next',1]])$r(k).onclick=()=>{const s=$r('image');if(s.options.length){s.selectedIndex=(s.selectedIndex+n+s.options.length)%s.options.length;open();}};
 for(const id of ['decision','reason','cluster','singleton'])$r(id).oninput=()=>{const iid=dataset.cases[currentCase].image_id,c=$r('mode').value,k=iid+'|'+c;if(!units.has(k))return;const x=$r('decision').value,v={status:['未审核','暂缓'].includes(x)?x:'已审核',grade:['未审核','暂缓'].includes(x)?'':x,reason:$r('reason').value,cluster_semantics:$r('cluster').value,singleton_assessment:$r('singleton').value};validate(v);edits[k]=v;save();};
 $r('note').oninput=()=>{notes[dataset.cases[currentCase].image_id]=$r('note').value;save();};
 $r('export').onclick=()=>{const v=exportData();validateImport(v);const url=URL.createObjectURL(new Blob([JSON.stringify(v,null,2)],{type:'application/json'})),a=document.createElement('a');a.href=url;a.download='图片历史难度_我的审核.json';a.click();setTimeout(()=>URL.revokeObjectURL(url),1000);};
 $r('import').onchange=async e=>{try{const f=e.target.files[0];if(!f)return;({edits,notes}=validateImport(JSON.parse(await f.text())));blocked=false;save();render();}catch(error){$r('save').textContent='导入失败，现有审核保留：'+error.message;}finally{e.target.value='';}};
 document.addEventListener('studio-case',render);$r('save').textContent=message;choices();render();
 window.DIFFICULTY_REVIEW_API={exportData,validateImport};
})();

File: tools/thesis_main/analysis/image_portrait/review_handoff.py
"""Attach current research candidates to the existing Panorama Studio.

No classification, coordinate repair, model inference or human adjudication.
The portable evidence excludes image bytes; the generated Studio stays local.
"""
from __future__ import annotations
import base64
import collections
import csv
import gzip
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[4]
B = ROOT / 'analysis_results/image_portrait_20260914_v1'
V1 = B / 'cloud/history_difficulty_20260915_v1/run_13859d59'
V2 = B / 'cloud/history_difficulty_review_20260915_v2/run_c0069628'
OUT = B / 'review_workflow_20260915'
LOCAL = B / 'local_review_studio'
OLD = ROOT / 'analysis_results/panorama_studio_20260907_v3'
STUDIO = ROOT / 'tools/label_studio/panorama_studio'
HERE = Path(__file__).resolve().parent


def read_rows(path):
    opener = gzip.open if path.suffix == '.gz' else open
    with opener(path, 'rt', encoding='utf-8-sig', newline='') as f:
        return list(csv.DictReader(f)) if '.csv' in path.name else [json.loads(x) for x in f if x.strip()]


def encoded(value):
    return json.dumps(value, ensure_ascii=False, allow_nan=False).replace('</', '<\\/')


def write_json(path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2, allow_nan=False) + '\n', encoding='utf-8')


def validate_review(v):
    if v.get('status') not in ('未审核', '已审核', '暂缓') or not isinstance(v.get('reason'), str):
        raise ValueError('invalid review status/reason')
    if v.get('grade') not in ('', '简单', '中等', '困难候选', '三档不适用'):
        raise ValueError('invalid reviewed grade')
    if (v['status'] == '已审核') != bool(v['grade']):
        raise ValueError('reviewed grade must be explicitly adjudicated')


def validate_memberships(groups, raw):
    seen, people = set(), set()
    for group in groups:
        counts = set()
        for cid in group:
            if cid in seen or raw[cid]['worker_id'] in people:
                raise ValueError('duplicate annotation/person in partition')
            seen.add(cid); people.add(raw[cid]['worker_id'])
            counts.add(len(raw[cid]['effective_points_1024x512']))
        if len(counts) > 1:
            raise ValueError('different point count in one cluster')


def run(focus=False):
    local = LOCAL / 'key39' if focus else LOCAL
    out = OUT / 'key39' if focus else OUT
    images = {r['image_id']: r for r in read_rows(B / 'metadata/images.jsonl')}
    raw = {r['canonical_annotation_id']: r for r in read_rows(B / 'human/responses.jsonl.gz')}
    tags = {r['image_id']: r for r in read_rows(V1 / 'expert/independent_tags106.csv')}
    first = {(r['image_id'], r['condition']): r for r in read_rows(V1 / 'targets/primary_with_robustness.csv')}
    latest = {(r['image_id'], r['条件']): r for r in read_rows(V2 / 'USER_DECISION_TABLE.csv')}
    assert len(latest) == len(first) == 230 and latest.keys() == first.keys()
    assert all(not r[k] for r in latest.values() for k in ('用户最终粗类', '用户裁决原因', '簇语义裁决', '单人标法裁决')), 'Preserve returned human edits before regenerating'
    historical = {i for i, _ in latest}; ids = historical | tags.keys()
    assert len(historical) == 205 and len(tags) == 106 and len(ids) == 276
    metadata = {r['image_id']: r for r in read_rows(V1 / 'inputs/image_metadata_whitelist.csv')}
    memberships = collections.defaultdict(lambda: collections.defaultdict(list))
    for r in read_rows(V2 / 'structure/real_mode_memberships.csv.gz'):
        if float(r['cut']) != .1:
            continue
        source = raw[r['canonical_annotation_id']]
        assert (source['image_id'], source['raw_condition'], source['worker_id']) == (r['image_id'], r['condition'], r['worker_id'])
        assert source['worker_id'] not in ('W019', 'W026')
        memberships[r['image_id'], r['condition']][r['cluster']].append(r['canonical_annotation_id'])
    for groups in memberships.values():
        validate_memberships(list(groups.values()), raw)
    queue = []
    if focus:
        study = B / 'cloud/image_links_after_review_20260915_v1/run_b02a97e2'
        queue = read_rows(study / 'local_image_review_queue.csv')
        ids = {r['image_id'] for r in queue}
        assert len(queue) == 44 and len(ids) == 39
        metadata.update({r['image_id']: r for r in read_rows(study / 'inputs/image_metadata_whitelist.csv')})
        for question in queue:
            for field in ('canonical_ids', 'representative_canonical_ids'):
                for cid in filter(None, question[field].split(';')):
                    assert raw[cid]['image_id'] == question['image_id'], (field, cid)
        for r in read_rows(study / 'oos/mode_memberships.csv'):
            source = raw[r['canonical_annotation_id']]
            assert (source['image_id'], source['worker_id'], source['raw_condition']) == (r['image_id'], r['worker_id'], 'oos')
            assert source['worker_id'] not in ('W019', 'W026')
            memberships[r['image_id'], 'oos_geometry'][r['cluster']].append(r['canonical_annotation_id'])
        for groups in memberships.values():
            validate_memberships(list(groups.values()), raw)

    # Reuse existing per-person 3D reconstructions only after checking exact raw points.
    old_text = (OLD / 'history_data.js').read_text(encoding='utf-8')
    old_cases, _ = json.JSONDecoder().raw_decode(old_text.split('cases.push(...', 1)[1])
    cached = {r['image_id']: r for r in old_cases}
    focused = {r['image_id'] for r in read_rows(V2 / 'expert/final_medium_hard_focus.csv')}
    ordered = sorted(ids, key=lambda i: (i not in focused, images[i]['building'], images[i]['number']))
    rows, cases, counts = [], [], collections.Counter()
    (local / 'history').mkdir(parents=True, exist_ok=True)
    for index, iid in enumerate(ordered):
        im = images[iid]; source_image = ROOT / im['path']
        assert source_image.is_file(), iid
        units = [dict(condition=c, previous=first[i, c], candidate=latest[i, c]) for i, c in latest if i == iid]
        row = dict(image_id=iid, building=im['building'], number=im['number'], image_path=im['path'],
                   scene=metadata[iid]['scene_category'], expert=tags.get(iid), units=units,
                   priority=iid in focused, original_image_checked=False,
                   questions=[dict(q, question_id=f'Q{j+1:02}') for j, q in enumerate(queue) if q['image_id'] == iid])
        rows.append(row)
        if iid in cached:
            text = (OLD / cached[iid]['history_script']).read_text(encoding='utf-8')
            payload, _ = json.JSONDecoder().raw_decode(text.split(',', 1)[1])
            photo, _ = json.JSONDecoder().raw_decode(text.split('const image=', 1)[1])
            variants = payload['variants']
            for v in variants:
                r = raw[v['source']['canonical_annotation_id']]
                assert r['image_id'] == iid and r['raw_points_1024x512'] == v['source']['raw_points']
                assert r['effective_points_1024x512'] == v['source']['effective_points']
                v['source']['main_worker_included'] = r['main_worker_included']
                if not r['main_worker_included']:
                    v['name'] += ' · 不纳入当前主分析'
                counts['reused_human_variants'] += 1
        else:
            assert iid not in historical, 'Historical geometry cache missing'
            photo = 'data:image/png;base64,' + base64.b64encode(source_image.read_bytes()).decode('ascii')
            variants = [dict(name='仅原图：没有本轮Manual/Semi历史', source={'role': 'image_only'}, error='没有真人布局；不生成候选难度或3D')]
        lookup = {v['source'].get('canonical_annotation_id'): j for j, v in enumerate(variants)}
        parts = {}
        for mode, condition in [('Manual', 'manual'), ('Semi', 'semi'), ('OOS', 'oos_geometry')]:
            groups = list(memberships.get((iid, condition), {}).values())
            parts[mode] = dict(status='第二轮候选分区；不是人工裁决',
                               candidates=[[[lookup[cid] for cid in g] for g in groups]] if groups else [])
            counts['clustered_responses'] += sum(map(len, groups))
        payload = dict(variants=variants, history=dict(partitions=parts), history_loaded=True)
        case = dict(image_id=iid, title=f"{im['building']} · {im['number']:02} · {row['scene']}",
                    category='图片难度审核 · 算法候选与人工结论分开', history_image=True,
                    history_script=f'history/{index}.js', variants=[])
        cases.append(case)
        script = f'Object.assign(window.STUDIO_DATA.cases[{index}],{encoded(payload)});\n'
        script += f'{{const image={encoded(photo)};window.STUDIO_IMAGES[{index}]={{original:image,texture:image}};}}'
        (local / case['history_script']).write_text(script, encoding='utf-8')
    if not focus:
        assert counts['clustered_responses'] == 2152
    portable = dict(schema='image_difficulty_review_evidence_v1', source_v1=str(V1.relative_to(ROOT)),
                    source_v2=str(V2.relative_to(ROOT)), early_k_candidates=list(range(2, 9)), rows=rows,
                    human_reviews=0, image_bytes_included=False, focus=focus,
                    image_root='../../../../' if focus else '../../../',
                    note='生成可看工具不等于实际看图；尚未审核不等于确认算法类别')
    write_json(out / 'review_evidence.json', portable)
    write_json(out / 'coverage.json', dict(images=len(rows), historical_images=len(ids & cached.keys()), image_conditions=sum(len(r['units']) for r in rows), questions=len(queue),
               original_tags=len(ids & tags.keys()), tag_overlap=len(ids & historical & tags.keys()), candidate_counts=dict(collections.Counter(
               r['条件'] + ':' + (r['建议粗类_非最终'] or '未分类') for (i, _), r in latest.items() if i in ids)),
               human_reviews=0, fresh_hd_image_reviews=0, **counts))
    groups = [{'code': 'REVIEW_ALL', 'building': '全部研究图片（不是同房）', 'images': [dict(id=r['image_id'], number=r['number'],
               manual=int(latest.get((r['image_id'], 'manual'), {}).get('原作答人数', 0)),
               semi=int(latest.get((r['image_id'], 'semi'), {}).get('原作答人数', 0))) for r in rows]}]
    dataset = dict(cases=cases, counts=dict(cases=len(cases), variants=counts['reused_human_variants']), annotation_writeback=False)
    (local / 'data.js').write_text('window.STUDIO_IMAGES={};window.STUDIO_DATA=' + encoded(dataset) + ';\nwindow.STUDIO_HISTORY=' + encoded(dict(groups=groups)) + ';\nwindow.DIFFICULTY_REVIEW=' + encoded(portable) + ';', encoding='utf-8')
    for name in ['studio.js', 'studio.css', 'history.css', 'index.html']:
        (local / name).write_bytes((STUDIO / name).read_bytes())
    for name in ['three.min.js', 'OrbitControls.js']:
        (local / name).write_bytes((STUDIO.parent / name).read_bytes())
    history = (STUDIO / 'history.js').read_text(encoding='utf-8')
    old_rule = 'q_boundary、q_wallwall均≥0.95，点数不同必分开；≥2人称簇，单人标法保留。使用人工复核后点；分簇不是收敛结论。'
    assert old_rule in history
    history = history.replace(old_rule, '本轮d_mask完整链接阈值0.10；点数不同分开，支持簇至少2位不同人员。点集叠加为已确认计算视图；3D复用原始顺序，二者可能不同。分簇不等于合理解释或收敛。')
    history = history.replace('同房组 · 历史标注', '当前研究图片 · 真实作答对照').replace('同房展示组', '图片集合')
    history = history.replace('<option>Semi</option>', '<option>Semi</option><option>OOS</option>')
    (local / 'history.js').write_text(history, encoding='utf-8')
    (local / 'review.js').write_bytes((HERE / 'review_handoff.js').read_bytes())
    page = (local / 'index.html').read_text(encoding='utf-8')
    page = page.replace('<script defer src="studio.js"></script>', '<script defer src="studio.js"></script><script defer src="history.js"></script><script defer src="review.js"></script><link rel="stylesheet" href="history.css">')
    (local / 'index.html').write_text(page, encoding='utf-8')
    print(json.dumps(dict(images=len(rows), units=sum(len(r['units']) for r in rows), questions=len(queue), **counts)), flush=True)


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--focus39', action='store_true')
    run(parser.parse_args().focus39)
